src/radical/pilot/agent/executing/popen.py

```diff
@@ -462,10 +534,226 @@ def _check_running(self):
+                    % os.environ['RP_APP_TUNNEL']
+
+        return ret
+
+
+    # --------------------------------------------------------------------------
+    # exec
+    #
+    def _get_task_env(self, task, launcher):
+
+        ret = ''
+        td  = task['description']
+
+        # named_env's are prepared by the launcher
+        if td['named_env']:
+            ret += '\n# named environment\n'
+            ret += '. %s\n' % launcher.get_task_env(td['named_env'])
+
+        # also add any env vars requested in the task description
+        if td['environment']:
+            ret += '\n# task env settings\n'
+            for key,val in td['environment'].items():
+                ret += 'export %s="%s"\n' % (key, val)
+
+        return ret
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _get_pre_exec(self, task):
+
+        ret  = ''
+
+        cmds = task['description']['pre_exec']
+        for cmd in cmds:
+            ret += '%s || (echo "pre_exec failed"; false) || exit 1\n' \
+                   % cmd
+
+        return ret
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _get_rank_ids(self, n_ranks, launcher):
+
+        ret  = ''
+        ret += 'export RP_RANKS=%s\n' % n_ranks
+        ret += launcher.get_rank_cmd()
+
+        return ret
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _get_post_exec(self, task):
+
+        ret  = ''
+        cmds = task['description']['post_exec']
+        for cmd in cmds:
+            ret += '%s || (echo "post_exec failed"; false) || exit 1\n' \
+                   % cmd
+
+        return ret
+
+
+    # --------------------------------------------------------------------------
+    #
+    # rank
+    #
+    def _get_pre_rank(self, rank_id, cmds=[]):
+
+        ret = ''
+        for cmd in cmds:
+            # FIXME: exit on error, but don't stall other ranks on sync
+            ret += '        %s\n' % cmd
+
+        return ret
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _get_rank_sync(self, sig):
+
+        # FIXME: make sure that all ranks are alive
+        ret  = 'while test $(cat %s.sig | wc -l) -lt $RP_RANKS; do\n' % sig
+        ret += '    sleep 1\n'
+        ret += 'done\n'
+
+        return ret
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _get_rank_exec(self, task, rank_id, rank, launcher):
+
+        # FIXME: this assumes that the rank has a `gpu_maps` and `core_maps`
+        #        with exactly one entry, corresponding to the rank process to be
+        #        started.
+
+        ret  = ''
+        gmap = rank['gpu_map'][0]
+        if gmap:
+            gpus = ','.join([str(gpu) for gpu in gmap])
+            ret += '        export CUDA_VISIBLE_DEVICES=%s\n' % gpus
```

Comment:
just a future reminder to check LMs, which have `cvd_id_mode = lm_info.get('cvd_id_mode', 'physical')`

p.s. from obsolete Scheduler's method `_handle_cuda(self, task)`

Reply:
right - I added a FIXME for now.